File: src/net/dinkla/raytracer/textures/SphericalMap.java
package net.dinkla.raytracer.textures;

import net.dinkla.raytracer.math.MathUtils;
import net.dinkla.raytracer.math.Point3D;
import net.dinkla.raytracer.utilities.Resolution;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 02.06.2010
 * Time: 21:13:41
 * To change this template use File | Settings | File Templates.
 */
public class SphericalMap extends Mapping {

    // TODO move to MathUtils or Constants
    final double 	invTWO_PI 	= 0.1591549430918953358;
    final double 	invPI 		= 0.3183098861837906715;

    @Override
    public Mapped getTexelCoordinates(Point3D p, Resolution res) {
        Mapped result = new Mapped();

        double theta =  Math.acos(p.getY());
        double phi =  Math.atan2(p.getX(), p.getZ());
        if (phi < 0) {
            phi += 2.0 * Math.PI;
        }

//        double u = phi * (1.0 / (2.0 *  Math.PI));
        double u =  (phi * invTWO_PI);
//        double v = 1 - theta * MathUtils.INV_PI;
        double v =  (1 - theta * invPI);

        result.column = (int) ((res.hres -1) * u);
        result.row = (int) ((res.vres -1) * v);

        return result;
    }
}

File: src/main/kotlin/net/dinkla/raytracer/textures/SphericalMap.kt
package net.dinkla.raytracer.textures

import net.dinkla.raytracer.math.MathUtils.PI
import net.dinkla.raytracer.math.MathUtils.INV_PI
import net.dinkla.raytracer.math.MathUtils.INV_TWO_PI
import net.dinkla.raytracer.math.Point3D
import net.dinkla.raytracer.utilities.Resolution
import kotlin.math.acos
import kotlin.math.atan2

class SphericalMap : Mapping() {

    override fun getTexelCoordinates(p: Point3D, res: Resolution): Mapping.Mapped {

        val theta = acos(p.y)
        var phi = atan2(p.x, p.z)
        if (phi < 0) {
            phi += 2.0 * PI
        }

        //        double u = phi * (1.0 / (2.0 *  Math.PI));
        val u = phi * INV_TWO_PI
        //        double v = 1 - theta * MathUtils.INV_PI;
        val v = 1 - theta * INV_PI

        val row = ((res.vres - 1) * v).toInt()
        val column = ((res.hres - 1) * u).toInt()

        return Mapped(row, column)
    }
}
